### PR TITLE
Additional Requirements

### PR DESCRIPTION
**Pull request by sudara**
Here's a first set of a wider array of requirements. I'm 50-50 on whether "owner" makes sense as a role, but I'm using it here for bigger-picture user-centric "guidance" type requirements.

### CODE REVIEW

File: docs/requirements.md [hunk 1]
```diff
@@ -1,16 +1,59 @@
-# PerfCheck CI Requirements
+# Perf Check CI Requirements
 
 Perf Check CI is a web application to manage jobs that measure and compare performance on another web application.
 
 ## Creating a job
 
-* As a user I want to create a job to compare the performance of two branches across multiple request paths.
+* As a user, I want to compare relative performance of my branch against master to make sure I'm not shipping code that will slow things down.
```
Comment (Manfred) [inline docs/requirements.md]: Do you think we should configure a fixed reference branch or are you using `master` as shorthand for the branch that is deployed to production?

File: docs/requirements.md [hunk 2]
```diff
@@ -1,16 +1,59 @@
...
+* As a user, I want to have a quick idea how performance on my branch looks, without caring about comparisons against master.
+* As a user, I want to create a job to compare the performance of two branches across multiple request paths.
```
Comment (Manfred) [inline docs/requirements.md]: What is the motivation to compare branches?

File: docs/requirements.md [hunk 3]
```diff
@@ -1,16 +1,59 @@
...
+* As a user, I know the request I'm working on takes a long time (10-60 seconds), so just want a very rough comparison that takes a smaller amount of time to complete so I can get back to work.
```
Comment (Manfred) [inline docs/requirements.md]: What is the pertinent different between this and `quick idea how performance on my branch looks`? Would this requirement exist if benchmarking would only take 5 seconds?

File: docs/requirements.md [hunk 4]
```diff
@@ -1,16 +1,59 @@
...
+* As a user, I care about getting a final and accurate-as-possible comparison across two branches.
```
Comment (Manfred) [inline docs/requirements.md]: Is this story about trust?

* As a user, I want Perf Check CI to return final, stable, and accurate comparisons between two branches so everyone involved in the project can understand and trust the result.

File: docs/requirements.md [hunk 5]
```diff
@@ -1,16 +1,59 @@
...
+
+* As a user, I want to know if the branch performance of each URL is better, about the same, or worse than the reference measurement.
+* As a user, I want to understand how the branch performance thresholds for better/worse are calculated and be able to verify and double check the calculation.
+* As a user, I want to know if URL(s) on my branch still have problematic performance in an absolute way, such as too many database queries, are too long for an acceptable web request, or eats up too much memory.
+
+## Workflow & CI
+
+* As a user, I want to easily start a job from a Github PR without many steps beyond knowing which URL(s) to test.
+* As a user, I want to have a small set of URLs that can be tested against master on most (but not all) of my team's PRs.
+* As a code reviewer, I want to see iterative history that performance was later improved on a PR that initially had poor performance.
```
Comment (Manfred) [inline docs/requirements.md]: * As a code reviewer, I want to see the history of Perf Checks for a branch so I can verify that the author made performance improvements I requested.

It might be good to be a bit more general here? A reviewer might want to see checks from before the PR was created because: there may be more information and a PR can get re-opened for the same branch. I also think we may want to be careful about suggesting anything about absolute performance in stories. It's possible that a branch already has good performance in the check but it shows that it's doing unexpected queries with caching turned on.

File: docs/requirements.md [hunk 6]
```diff
@@ -1,16 +1,59 @@
...
+
+## Social Proof
+
+* As a user, I want my GitHub PR to display an up-to-date concise summary of performance status that links to details of the latest Perf Check job for the branch.
```
Comment (Manfred) [inline docs/requirements.md]: Maybe we need an additional first story to set the motivation for all additional stories. Something like:

* As a user, I want to share Perf Checks so I can prove performance details.

File: docs/requirements.md [hunk 7]
```diff
@@ -1,16 +1,59 @@
...
+* As a user, I'd like to share my results summary with someone in slack.
+
+## Error Handling
+
+* As a user, I want to know if everything on Perf Check CI is actually working right now.
```
Comment (Manfred) [inline docs/requirements.md]: I think it makes sense to include some motivation here: *so I can verify that the service is operating properly when my Checks are failing*.

File: docs/requirements.md [hunk 8]
```diff
@@ -1,16 +1,59 @@
...
+* As a Perf Check developer, I want to know if all specific components of the Perf Check CI system are functioning properly (db online, db last updated, queue status, target app bootable).
+* As the product owner, I want to communicate clearly to a user when a job failure or error was the result of problems with their branch or user supplied Job settings.
```
Comment (Manfred) [inline docs/requirements.md]: … so I can help users fix their own failed Checks?

File: docs/requirements.md [hunk 9]
```diff
@@ -1,16 +1,59 @@
...
 ## Measurements and statistics
 
-* As a developer I want to store all measurements from a PerfCheck run so can process them later.
-* As a developer I want to compute summary statistics over measurements so I can use them for reporting.
+* As a product owner, I want to present very limited and clear "headline" guidance to users so they attend to and optimize "big picture" performance instead of being distracted by Perf Check implementation details.
+* As a Perf Check developer, I want to store all measurements from a Perf Check run so can process them later.
+* As a Perf Check developer, I want to compute summary statistics over measurements so I can use them for reporting.
+* As a user, I want to see a detailed breakdown of each request made.
+* As a user, I want to know variance of a group of requests and know how "normal" that variance is, so I feel comfortable trusting the summarized results and knowing whether the request profile is changed or unchanged.
+* As a user, I want to gain intuition about request performance variance via a visual representation of where the reference and branch requests fall on a shared timeline.
```
Comment (Manfred) [inline docs/requirements.md]: Has anyone you spoke to actually voiced concerns over variance in the measurements? Should we summarize the last three stories?

* As a user, I want to see all benchmarked requests with an explanation of how they influence the Perf Check result so I can have more trust in the algorithm.

**Comment by Manfred** [comment]
Can we choose a short list of consistent words or phrases for certain things across the stories?

* Perf Check / Job / Check? What do we call it?
* URL / Request path? If we choose URL, we have to define that a URL is always on the target application and not something that can be filled out randomly.
* Master / mainline branch / reference branch? I'm fine with using master as long as there is a definition at the end of the stories that we use *master* as a shorthand for the deployed branch.

File: docs/requirements.md (hunk 1 above)
Comment (tvandervossen) [reply docs/requirements.md]: Consistently using “reference branch” might make sense as you also use “reference measurement” to refer to the results of this branch further down in these requirements.

Comment (sudara) [reply docs/requirements.md]: Yes, I was also wondering what to do about about consistency issues. I voiced these drafts from the point of view of the user, who 95% of the time will only care about master and only in edge cases will care that the reference branch can be other branches. 

I’m fine with altering them to be consistent if that’s preferred over an “accurate” user point of view. *1 for “Reference Branch.”

In terms of defaults in the UI, we should absolutely be assuming they want to reference against master.

File: docs/requirements.md (hunk 2 above)
Comment (sudara) [reply docs/requirements.md]: I’ll rephrase this to make it clear it’s specifying non-master/arbitrary  reference branches. 

Its most relevant when having the reference branch be a feature branch and the test branch being a branch with an alternative implementation, for example one that includes perf fixes. It’s less important than master/production branch.

File: docs/requirements.md (hunk 3 above)
Comment (sudara) [reply docs/requirements.md]: The “quick idea” is not comparative. It’s test branch only and used for iterating, usually when you already know the values for master/reference. 

The “long request” version is comparative but for example instead of running 20 requests on both test/reference, it’s 5-10. 

It’s essentially a way to specify ‘n’ should be adaptive without specifying that a user needs to directly specify the actual quantity.

File: docs/requirements.md (hunk 4 above)
Comment (sudara) [reply docs/requirements.md]: This is a way of specifying that people might want to increase ‘n’ for increased accuracy, for example after arriving at a final result after optimization.

File: docs/requirements.md (hunk 5 above)
Comment (sudara) [reply docs/requirements.md]: This was poorly phrased. I actually meant within a PR. I’ll rethink. 

The specific case is about workflow re: a branch that initially showed performance problems/regressions. The rest of the team is now aware problems exist and would appreciate seeing the iteration towards a more acceptable result. 

It’s not about pre/post PR perf checks or some source of truth about the start of a branch. It’s about opting in to iterative improvement attempts, preferably within the PR. 

Good point on multi-PRs, but not sure we “know” anything about whether we should handle that case yet.

File: docs/requirements.md (hunk 9 above)
Comment (sudara) [reply docs/requirements.md]: I'm fine with conflating the last two. Seeing the individual requests seems important, as the CLI provides this and removing it would be confusing, I think.

**Comment by sudara** [comment]
@manfred Yes, some consistency would be good. My votes are:

* Job (for something that a user kicks off) might be clearer than "Perf Check" which is currently being used as a qualifier for "developer" (which is unclear on its own) and the name of both the projects.
* The Application for "Perf Check CI"
* Perf Check CLI for the command line tool.
* Request Path is what's currently in the UI. Since the users are actually mostly developers, it's going to reduce confusion to stick with that.
* I'd prefer to use "master" or the less clear "default reference branch" to specifically represent "the branch currently deployed to production that most people want to compare with most of the time" and otherwise "reference branch" for 1) the feature/edge case where the reference branch is set to something else or 2) when we generally speak about comparisons.

File: docs/requirements.md (hunk 9 above)
Comment (sudara) [reply docs/requirements.md]: Rethinking this, I think variance is probably only really important to Perf Check developers. Presenting it to users might be friendly, assuming the context is provided (i.e. how normal the spread is). I'll come up with something that tries to be generic.

Comment (sudara) [reply docs/requirements.md]: Ah there is overlap here with 

* As a user, I want to understand how the branch performance thresholds for better/worse are calculated and be able to verify and double check the calculation.

File: docs/requirements.md (hunk 3 above)
Comment (sudara) [reply docs/requirements.md]: I think I "see" your main point. I'm updating language, but keeping the two separate as they are different use cases and different ways to cut time down. If everything took 5 seconds, we wouldn't have to build this tool at all :D Feel free to push back though.

File: docs/requirements.md (hunk 5 above)
Comment (sudara) [reply docs/requirements.md]: > I also think we may want to be careful about suggesting anything about absolute performance in stories. It's possible that a branch already has good performance in the check but it shows that it's doing unexpected queries with caching turned on.

Maybe we can pull this conversation out of this PR. Not entirely sure what you mean here.

File: docs/requirements.md (hunk 1 above)
Comment (Manfred) [reply docs/requirements.md]: Requirements are written from the perspective of the user or role by that doesn't mean it has to be in their voice or wording. I'm fine whatever we use as long as it makes the most sense or is usable and consistent and we all know what it means.

Wording in the requirements don't have to appear in the UI. However, if you're lucky you can have consistent naming across stories, implementation, and UI.

Comment (sudara) [reply docs/requirements.md]: >  doesn't mean it has to be in their voice or wording.

Yeah, this tripped me up a bit generally. A user isn't always going to know feature ABC or domain language XYZ. It seems strange to have requirements about the user written from the developer point of view and knowledge — we could easily misinterpret / misconstrue / fail to identify what the actual user point of view and task is — which I thought was part of the point of the exercise.

**Comment by sudara** [comment]
It would be helpful if either of you two could point out specific useful clarification or understanding that resulted from writing these (or if you expect them to be particularly useful to in the future). 

Meaning, I understand the *theoretical* value of requirements.md but I would appreciate knowing when and where you two get *actual* value from it in the context of this project so I can better understand and judge its role. 

My guess is:

1. @tvandervossen might have not been aware of all of these, so it's been useful to iterate through them.
2.  @manfred and I may sometimes have different angles that we see features from, so being specific/careful about wording is a proxy for/excuse to preemptively shake out initial expectations. I typically expect direct conversations around any feature building, as needs are ever shifting and the problems are three dimensional, not a flat list. But perhaps this document has given us the chance to shake out a few initial expectations, even if vague and indirect.
3. We can update it regularly to be a source of truth for <reasons> (not sure what other reasons may pertain to perf check ci beyond the above two). 
4. This doc becomes a living document containing post-decision rationalization. Moving forward, it provides some similar confidence/momentum to team members as a roadmap might.
5. The intention is to treat items as a "implicit soft green light" for feature development. Anything on this list is forrmally "fair game" for someone to run ahead with and start chewing on. Likewise, anything not on this list needs to get on this list before we are "officially" building it. I doubt you guys will agree this is the case, but wondering if there's some truth to it.

File: docs/requirements.md (hunk 1 above)
Comment (tvandervossen) [reply docs/requirements.md]: >  It seems strange to have requirements about the user written from the developer point of view and knowledge

Why? User stories aren’t intended to be autobiographical. In most cases, the user’s perspective or motivation can be expressed more clearly and more precisely using language or knowledge this user might lack.

>  we could easily misinterpret / misconstrue / fail to identify what the actual user point of view and task is

In my experience, clear and consistent terminology in the requirements helps rather than hinder this.

**Comment by Manfred** [comment]
Yes, mostly like you said.

It's useful for people who are new or not up-to-date with the project to read what it's supposed to do and why. Having one document explain the entire application is much easier than reading all the code or previous issues and pull requests. Understanding how a certain feature works in the context of others gives useful perspective when working on the implementation. We've also used it in the past to talk to security and privacy officers to help with auditing and compliance.

Wording the requirements carefully allows us to iron out as many misunderstandings as possible. We can use the requirements to determine priorities and focus on a certain part of the implementation: *we're doing 12 before 15* or *that's 54, we're currently only working on 26*.

I don't think it's a green light to just start developing because you still want to prioritize and design before you start typing code or creating pictures or whatever. These can be tiny steps or just a doodle on a piece of paper, but I think that's still important.

File: docs/requirements.md (hunk 1 above)
Comment (sudara) [reply docs/requirements.md]: > Why?

My intuition is that we often work with a different set of assumptions and language than a user might, which puts as at risk of not properly addressing their assumptions and instead rolling with our own. By striving to keep language compatible and consistent with user expectations, we might better understand and solve for their requirement.

> clear and consistent terminology in the requirements helps

100% with you on this. There are always multiple ways to be consistent, each with their tradeoffs. Just trying to clarify which tradeoffs you prefer...

**Comment by sudara** [comment]
@manfred Thanks! I appreciate the clarification on the "green light." Rephrasing so that I'm clear on how you expect these to help with this project in particular:

* They will help you prioritize when building by roughly defining scope
* They provide you with a larger context when working on a requirement, preventing misunderstandings
* They will help anyone out of date / a new person get up to speed

Thanks for letting me dig in. I work completely differently than this, so it's useful to know your preferences and expectations.

File: docs/requirements.md (hunk 1 above)
Comment (tvandervossen) [reply docs/requirements.md]: As Manfred mentioned, it’s of course nice to have consistent naming across stories, implementation, and UI.

I think we’re all experienced enough to be able to translate in cases where we feel it makes sense to use more “formal” or more “developer-y” terminology in our requirements or in the source code, while we’re using more “informal” or longer, more explanatory UI copywriting for the user.

In this specific case, the other option is to consistently use the term “master” instead of “reference branch” in the requirements with a short note or definition below the user stories that “master” is used as a shorthand for “reference branch” and that this might be something else in some cases as Manfred mentions above.

As with a lot of things, it’s a matter of finding a balance that makes sense. :)

Comment (sudara) [reply docs/requirements.md]: Sounds good! 

I updated this example to keep the distinction in place, as it specifically for the "default reference branch" (most clearly expressed as "master") and updated other usages where it's not relevant to the more generic "reference branch."

**Comment by Manfred** [comment]
I'm going to merge this and then edit it.
